### DISCUSSION TITLE
Retrieving string descriptors

### DISCUSSION THREAD

**Discussion by csBlueChip** [hathach/tinyusb#1125] [Q&A]
I am trying to read the string descriptor for the device `iProduct`/`iManufacturer`/`iSerialNumber`.

I have added an API call to expose these values : https://github.com/csBlueChip/tinyusb/commit/efe6fa88750f9e007216d651afa0427d762e4dcd

The first thing I need to do is read `Descriptor/String[0]` (to get the language table) ...So here https://github.com/csBlueChip/tinyusb/blob/bc-midi/src/class/midi/midi_host.c#L135 (inside `midih_set_config()`) I have added:
```
//                                      req  getdesc  idx   str     lang      datalen
TU_ASSERT( tuh_control_xfer(0,
  (tusb_control_request_t*)((uint8_t[]){0x80, 0x06,  0x00,  0x03, 0x00,0x00, 0xFE,0x00}),
  buf, strHandler) );  // get language table
```

I am getting back 254 `0x00`'s (rather than anything resembling a language table) ...and `strHandler()` is NOT being called.

I added some debug to `usbh_control.c` so I could check the function pointer was making it through https://github.com/csBlueChip/tinyusb/blob/bc-midi/src/host/usbh_control.c#L83 and found something I simply cannot understand ...The printf() _before_ the callback() does NOT happen, but the print() _after_ the callback() DOES happen.

Can someone please help me work out what is happening and how to fix it

I've got a full MIDI engine waiting to be debugged, and no way to talk to the USB<->MIDI cable ...So there's a potentially big win to be had if we can work this out.

**Comment by hathach** [comment]
just skip it for now, it isn't needed to get midi driver to work. I will add an dedicated API for getting string descriptor later on.

**Comment by csBlueChip** [comment]
Good point! 

Beyond the MIDI driver, I have a specific use-case where I have two devices which can only be identified by their Product string! <facepalm> 

However, for now, you are right, I need to ditch the RP2040 [because no/limited Bulk Endpoint support :( ] and get MIDI working before String Descriptors becomes important.

**Comment by rppicomidi** [comment]
@hathach I have an API working for retrieving string descriptors (I need it for a MIDI host application I am working on). Do you want a separate pull request or should I just commit it in my midihost pull request? This is the API:

```
//--------------------------------------------------------------------+
// USBH Device Strings API
//--------------------------------------------------------------------+
#ifndef CFG_TUH_DEVSTRINGS
#define CFG_TUH_DEVSTRINGS 0
#endif
// To use this API:
// Add "#define CFG_TUH_DEVSTRINGS 1" to tusb_config.h
// Implement your main task loop as a state machine with an initial
// "Disconnected" state.
// When the device connects, the application runs changes state machine state
// to a state that calls usbh_desc_istrings_get() to get the product string
// index numbers, then calls usbh_request_langids(). The function usbh_langids_cb()
// records the supported
// language IDs, chooses the appropriate language from the list, switches the
// state to "Getting Strings" and calls usbh_request_device_string() for the
// first string of interest (usually manufacturer or product). The usbh_dev_string_cb()
// callback function records the retrieved string and either calls 
// usbh_request_device_string() with another language ID or string index, or
// sets the state to the the normal operating state.
// When the device disconnects, the operating state changes back to "Disconnected"
// so if a new device connects, the application can fetch the strings for the new device.
#if CFG_TUH_DEVSTRINGS
// Get the string index numbers for the manufacturer, product and serial number
// that were read during enumeration.
bool usbh_desc_istrings_get(uint8_t dev_addr, uint8_t* i_manufacturer, uint8_t* i_product, uint8_t* i_serial);

// Request the language IDs. Call usbh_langids_cb() when done
bool usbh_request_langids(uint8_t dev_addr);

// Required application callback function (application must implement)
// Pass the number of language IDs and a pointer to an array that contains them
// The application must copy them because they are not available after this callback
// unless usbh_request_langids() is called again
void usbh_langids_cb(uint8_t dev_addr, uint8_t num_langids, uint16_t* langids);

// Request the device string with index istring and language ID langid
// Call usbh_dev_string_cb() when done
bool usbh_request_device_string(uint8_t dev_addr, uint8_t istring, uint16_t langid);

// Required application callback function (application must implement)
// Pass the the number of UTF16-LE 16-bit values in the device string
// and a pointer to the utf16le device string. The application must copy
// the utf16le string because it is not available after this callback unless
// the application calls usbh_request_device_string() again with the same istring
// and langid
void usbh_dev_string_cb(uint8_t dev_addr, uint8_t istring, uint16_t langid, uint8_t num_utf16le, uint16_t* utf16le);
#endif
```
```

**Comment by hathach** [comment]
Host string (along with other) descriptors API is added in #1369 

https://github.com/hathach/tinyusb/blob/4c6bb161e2a39774f0f63a6e230627506c1a4682/src/host/usbh.h#L92-L113
